File: dev-page/src/main/java/org/example/page/task/expert/user/signup/Controller.java
package org.example.page.task.expert.user.signup;

import org.example.core.page.presentation.AbstractController;
import org.example.intent.SignIn;

import javax.swing.*;
import java.awt.*;

class Controller extends AbstractController {
    private final Service service;

    Controller(Service service) {
        this.service = service;
    }

    @Override
    public JPanel getLayout() {
        JPanel layout = new JPanel();
        layout.setLayout(new BoxLayout(layout, BoxLayout.Y_AXIS));
        layout.setBorder(BorderFactory.createEmptyBorder(20, 20, 20, 20));

        JTextField idField = new JTextField(20);
        idField.setMaximumSize(new Dimension(Integer.MAX_VALUE, 30));
        idField.setToolTipText("ID");

        JPasswordField passwordField = new JPasswordField(20);
        passwordField.setMaximumSize(new Dimension(Integer.MAX_VALUE, 30));

        JButton signUpButton = new JButton("Sign Up");
        JButton homeButton = new JButton("Home");

        signUpButton.addActionListener(e -> onSignUpButtonClick(idField.getText(), new String(passwordField.getPassword())));
        homeButton.addActionListener(e -> onHomeButtonClick());

        layout.add(new JLabel("ID:"));
        layout.add(idField);
        layout.add(Box.createVerticalStrut(10));
        layout.add(new JLabel("Password:"));
        layout.add(passwordField);
        layout.add(Box.createVerticalStrut(20));
        layout.add(signUpButton);
        layout.add(Box.createVerticalStrut(10));
        layout.add(homeButton);

        return layout;
    }

    private void onSignUpButtonClick(String id, String pw) {
        if (id.isEmpty() || pw.isEmpty()) {
            alert();
            return;
        }

        service.signUp(id, pw);
        System.out.printf("Expert User Sign up. ID: %s, pw: %s%n", id, pw);
        dispatcher.dispatch(new SignIn());
    }

    private void alert() {
        JOptionPane.showMessageDialog(
                null,
                "Please fill out all fields.",
                "Sign Up Error",
                JOptionPane.WARNING_MESSAGE
        );
    }

    private void onHomeButtonClick() {
        dispatcher.dispatch(new SignIn());
        System.out.println("Go Home");
    }
}

File: dev-page/src/main/kotlin/org/example/page/task/expert/user/signup/Controller.kt
package org.example.page.task.expert.user.signup

import org.example.core.page.presentation.AbstractController
import org.example.intent.SignIn
import org.example.page.task.utils.*
import javax.swing.*

internal class Controller(
    private val expertUserService: Service
) : AbstractController(expertUserService) {

    override val layout: JPanel
        get() = panel {

            label("ID:")
            val idField = textField(tooltip = "ID")

            vSpace(10)

            label("Password:")
            val passwordField = passwordField()

            vSpace(20)

            button("Sign Up") {
                onSignUpButtonClick(
                    idField.text,
                    String(passwordField.password)
                )
            }

            vSpace(10)

            button("Home") {
                onHomeButtonClick()
            }
        }

    private fun onSignUpButtonClick(id: String, pw: String) {
        if (id.isBlank() || pw.isBlank()) {
            alert()
            return
        }

        expertUserService.signUp(id, pw)
        println("Expert User Sign up. ID: $id, pw: $pw")

        dispatcher?.dispatch(SignIn("SignIn"))
    }

    private fun alert() {
        JOptionPane.showMessageDialog(
            null,
            "Please fill out all fields.",
            "Sign Up Error",
            JOptionPane.WARNING_MESSAGE
        )
    }

    private fun onHomeButtonClick() {
        dispatcher?.dispatch(SignIn("SignIn"))
        println("Go Home")
    }
}
